### PR TITLE
[ALLUXIO-2804] Enable concurrent create paths

### PR DESCRIPTION
**Pull request by gpang**
https://alluxio.atlassian.net/browse/ALLUXIO-2804

During the implementation, the creation journaling was improved, to address: https://alluxio.atlassian.net/browse/ALLUXIO-2737

### CODE REVIEW

File: core/server/common/src/main/java/alluxio/master/journal/JournalContext.java [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+   * @param entry the {@link Journal.JournalEntry} to append to the journal
+   */
+  void append(Journal.JournalEntry entry);
+
+  /**
+   * @return the journal flush counter
+   */
+  long getFlushCounter();
+
+  @Override
```
Comment (jsimsa) [inline core/server/common/src/main/java/alluxio/master/journal/JournalContext.java]: this is not needed

File: core/server/master/src/main/java/alluxio/master/file/meta/Inode.java [hunk 2]
```diff
@@ -132,6 +132,23 @@ public PersistenceState getPersistenceState() {
   }
 
   /**
+   * Compare-and-swaps the persistence state.
+   *
+   * @param old the old {@link PersistenceState}
+   * @param newState the new {@link PersistenceState} to set
+   * @return true if the old state matches and the new state was set
+   */
+  public boolean compareAndSwapPersistenceState(PersistenceState old, PersistenceState newState) {
```
Comment (jsimsa) [inline core/server/master/src/main/java/alluxio/master/file/meta/Inode.java]: nit: you can remove the `PersistenceState` as the suffix of the method

File: core/server/master/src/main/java/alluxio/master/file/FileSystemMaster.java [hunk 3]
```diff
@@ -210,24 +211,26 @@
...
+   * "AndJournal".
    * (B) cannot call (A)
    * (B) can call (B)
    * (B) can call (C)
    * (B) cannot call (D)
    *
    * (C) private (or package private) internal methods:
-   * These methods perform the rest of the work, and do not do any journaling. The names of these
-   * methods are suffixed by "Internal".
+   * These methods perform the rest of the work, and may append to the journal. The names of these
```
Comment (jsimsa) [inline core/server/master/src/main/java/alluxio/master/file/FileSystemMaster.java]: OOC, why is it necessary to allow the "Internal" methods to append to the journal?

File: core/server/master/src/main/java/alluxio/master/file/FileSystemMaster.java [hunk 4]
```diff
@@ -739,20 +742,13 @@ public FileInfo getFileInfo(AlluxioURI path)
...
+        LockedInodePath inodePath = mInodeTree.lockInodePath(path, InodeTree.LockMode.READ)) {
       mPermissionChecker.checkPermission(Mode.Bits.READ, inodePath);
       if (inodePath.fullPathExists()) {
         // The file already exists, so metadata does not need to be loaded.
         return getFileInfoInternal(inodePath);
       }
-    }
-
-    try (JournalContext journalContext = createJournalContext();
-        LockedInodePath inodePath = mInodeTree.lockInodePath(path, InodeTree.LockMode.WRITE)) {
```
Comment (jsimsa) [inline core/server/master/src/main/java/alluxio/master/file/FileSystemMaster.java]: why is write lock no longer needed?

File: core/server/master/src/main/java/alluxio/master/file/FileSystemMaster.java [hunk 5]
```diff
@@ -829,8 +825,7 @@ public PersistenceState getPersistenceState(long fileId) throws FileDoesNotExist
       throws AccessControlException, FileDoesNotExistException, InvalidPathException {
     Metrics.GET_FILE_INFO_OPS.inc();
     try (JournalContext journalContext = createJournalContext();
-        LockedInodePath inodePath = mInodeTree.lockInodePath(path, InodeTree.LockMode.WRITE)) {
-      // This is WRITE locked, since loading metadata is possible.
+        LockedInodePath inodePath = mInodeTree.lockInodePath(path, InodeTree.LockMode.READ)) {
```
Comment (jsimsa) [inline core/server/master/src/main/java/alluxio/master/file/FileSystemMaster.java]: why is write lock no longer needed?

File: core/server/master/src/main/java/alluxio/master/file/meta/LockedInodePath.java [hunk 6]
```diff
@@ -115,6 +119,13 @@ public synchronized boolean fullPathExists() {
     return mInodes.size() == mPathComponents.length;
   }
 
+  /**
+   * @return the {@link alluxio.master.file.meta.InodeTree.LockMode} of this path
```
Comment (jsimsa) [inline core/server/master/src/main/java/alluxio/master/file/meta/LockedInodePath.java]: nit: no need for `alluxio.master.file.meta.`

File: core/server/master/src/main/java/alluxio/master/file/FileSystemMaster.java (hunk 4 above)
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/file/FileSystemMaster.java]: `getFileInfo` is a read operation, and the reason we used the WRITE lock before was to allow loading metadata. With the new creation locking scheme, we no longer have to use the WRITE lock mode, since the new scheme will take care of creating inodes along the path with the correct lock modes.

File: core/server/master/src/main/java/alluxio/master/file/FileSystemMaster.java (hunk 5 above)
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/file/FileSystemMaster.java]: Like the earlier response. `listStatus ` is a read operation, and the reason we used the WRITE lock before was to allow loading metadata. Now, the new locking scheme will handle creating inodes.

File: core/server/master/src/main/java/alluxio/master/file/meta/Inode.java (hunk 2 above)
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/file/meta/Inode.java]: done

File: core/server/master/src/main/java/alluxio/master/file/meta/LockedInodePath.java (hunk 6 above)
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/file/meta/LockedInodePath.java]: done

File: core/server/master/src/main/java/alluxio/master/file/FileSystemMaster.java (hunk 3 above)
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/file/FileSystemMaster.java]: This is related to the fact that we could be journaling while holding the READ lock. Before, holding the WRITE lock allowed us to treat it as a big lock to protect multiple operations, so we could journal at any time while holding the WRITE lock.

Now that we update state while holding the READ lock, journaling operations should be appended when they happen, because other threads could view the mutated state. Therefore, there needs to be more fine-grained journal appending, so internal methods may need to journal along the way.

We still disallow `fromEntry` methods from journaling, by using the `NoopJournalContext` in those methods.

File: core/server/common/src/main/java/alluxio/master/AbstractMaster.java [hunk 7]
```diff
@@ -303,31 +304,44 @@ protected ExecutorService getExecutorService() {
...
-    private long getFlushCounter() {
+    @Override
+    public long getFlushCounter() {
       return mFlushCounter;
     }
 
-    private void setFlushCounter(long counter) {
-      mFlushCounter = counter;
+    @Override
+    public void append(alluxio.proto.journal.Journal.JournalEntry entry) {
```
Comment (jsimsa) [inline core/server/common/src/main/java/alluxio/master/AbstractMaster.java]: nit: this can be just `append(JournalEntry entry)`

File: core/server/common/src/main/java/alluxio/master/journal/JournalContext.java [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+import java.io.Closeable;
+
+/**
+ * Context for storing journaling information.
+ */
+public interface JournalContext extends Closeable {
+  /**
+   * @param entry the {@link Journal.JournalEntry} to append to the journal
+   */
+  void append(Journal.JournalEntry entry);
```
Comment (jsimsa) [inline core/server/common/src/main/java/alluxio/master/journal/JournalContext.java]: nit: this can be just `append(JournalEntry entry)` if you import `alluxio.proto.journal.Journal.JournalEntry`

File: core/server/common/src/main/java/alluxio/master/journal/JournalContext.java [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+import alluxio.proto.journal.Journal;
+
+import java.io.Closeable;
+
+/**
+ * Context for storing journaling information.
+ */
+public interface JournalContext extends Closeable {
+  /**
+   * @param entry the {@link Journal.JournalEntry} to append to the journal
```
Comment (jsimsa) [inline core/server/common/src/main/java/alluxio/master/journal/JournalContext.java]: ditto

File: core/server/common/src/main/java/alluxio/master/journal/NoopJournalContext.java [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+
+  /**
+   * Constructs the {@link NoopJournalContext}.
+   */
+  public NoopJournalContext() {
+    // Do nothing
+  }
+
+  @Override
+  public void append(Journal.JournalEntry entry) {
```
Comment (jsimsa) [inline core/server/common/src/main/java/alluxio/master/journal/NoopJournalContext.java]: ditto

File: core/server/master/src/main/java/alluxio/master/block/BlockMaster.java [hunk 11]
```diff
@@ -32,6 +32,7 @@
 import alluxio.master.block.meta.MasterBlockInfo;
 import alluxio.master.block.meta.MasterBlockLocation;
 import alluxio.master.block.meta.MasterWorkerInfo;
+import alluxio.master.journal.JournalContext;
```
Comment (jsimsa) [inline core/server/master/src/main/java/alluxio/master/block/BlockMaster.java]: was this import missing before?

File: core/server/master/src/main/java/alluxio/master/file/FileSystemMaster.java (hunk 3 above)
Comment (jsimsa) [inline core/server/master/src/main/java/alluxio/master/file/FileSystemMaster.java]: I see ... I think it is confusing to keep to distinction between `(B)` and `(C)` if both methods can journal ... what if any is the logical difference between those two types of methods?

File: core/server/master/src/main/java/alluxio/master/file/FileSystemMaster.java (hunk 4 above)
Comment (jsimsa) [inline core/server/master/src/main/java/alluxio/master/file/FileSystemMaster.java]: thanks for the explanation

File: core/server/master/src/main/java/alluxio/master/file/FileSystemMaster.java (hunk 3 above)
Comment (jsimsa) [inline core/server/master/src/main/java/alluxio/master/file/FileSystemMaster.java]: would it make sense to inline `(B)` in `(A)`?

File: core/server/master/src/main/java/alluxio/master/file/FileSystemMaster.java [hunk 12]
```diff
@@ -1143,16 +1138,17 @@ private void createFileAndJournal(LockedInodePath inodePath, CreateFileOptions o
    *         option is false
    */
   InodeTree.CreatePathResult createFileInternal(LockedInodePath inodePath,
-      CreateFileOptions options)
+      CreateFileOptions options, JournalContext journalContext)
       throws InvalidPathException, FileAlreadyExistsException, BlockInfoException, IOException,
       FileDoesNotExistException {
-    InodeTree.CreatePathResult createResult = mInodeTree.createPath(inodePath, options);
+    if (mWhitelist.inList(inodePath.getUri().toString())) {
```
Comment (jsimsa) [inline core/server/master/src/main/java/alluxio/master/file/FileSystemMaster.java]: I assume this is a bug fix, correct?

File: core/server/master/src/main/java/alluxio/master/file/FileSystemMaster.java [hunk 13]
```diff
@@ -1311,20 +1304,21 @@ private void deleteFromEntry(DeleteFileEntry entry) {
 
   /**
    * Convenience method for avoiding {@link DirectoryNotEmptyException} when calling
-   * {@link #deleteInternal(LockedInodePath, boolean, long, DeleteOptions)}.
+   * {@link #deleteInternal(LockedInodePath, JournalContext, boolean, long, DeleteOptions)}.
    *
    * @param inodePath the {@link LockedInodePath} to delete
+   * @param journalContext the journal context
```
Comment (jsimsa) [inline core/server/master/src/main/java/alluxio/master/file/FileSystemMaster.java]: put the output argument after the input arguments

File: core/server/master/src/main/java/alluxio/master/file/FileSystemMaster.java [hunk 14]
```diff
@@ -1335,6 +1329,7 @@ private void deleteRecursiveInternal(LockedInodePath inodePath, boolean replayed
    * Implements file deletion.
    *
    * @param inodePath the file {@link LockedInodePath}
+   * @param journalContext the journal context
```
Comment (jsimsa) [inline core/server/master/src/main/java/alluxio/master/file/FileSystemMaster.java]: ditto

File: core/server/master/src/main/java/alluxio/master/file/FileSystemMaster.java [hunk 15]
```diff
@@ -1875,14 +1841,15 @@ private void renameAndJournal(LockedInodePath srcInodePath, LockedInodePath dstI
    *
    * @param srcInodePath the path of the rename source
    * @param dstInodePath the path to the rename destination
+   * @param journalContext the journal context
```
Comment (jsimsa) [inline core/server/master/src/main/java/alluxio/master/file/FileSystemMaster.java]: put the output argument after the input arguments

File: core/server/master/src/main/java/alluxio/master/file/FileSystemMaster.java [hunk 16]
```diff
@@ -1920,33 +1887,31 @@ private void renameInternal(LockedInodePath srcInodePath, LockedInodePath dstIno
...
+
+        // inodes from the same mount point as the dst
+        Stack<InodeDirectory> sameMountDirs = new Stack<>();
         List<Inode<?>> dstInodeList = dstInodePath.getInodeList();
-        Stack<Pair<String, MkdirsOptions>> ufsDirsToMakeWithOptions = new Stack<>();
-        AlluxioURI curUfsDirPath = new AlluxioURI(ufsDstUri).getParent();
-        // The dst inode does not exist yet, so the last inode in the list is the existing parent.
         for (int i = dstInodeList.size() - 1; i >= 0; i--) {
-          if (ufs.isDirectory(curUfsDirPath.toString())) {
+          // Since dstInodePath is guaranteed to not be a full path, all inodes in the incomplete
```
Comment (jsimsa) [inline core/server/master/src/main/java/alluxio/master/file/FileSystemMaster.java]: `... guaranteed not to be ...`

File: core/server/master/src/main/java/alluxio/master/file/meta/Inode.java [hunk 17]
```diff
@@ -132,6 +132,23 @@ public PersistenceState getPersistenceState() {
   }
 
   /**
+   * Compare-and-swaps the persistence state.
+   *
+   * @param old the old {@link PersistenceState}
```
Comment (jsimsa) [inline core/server/master/src/main/java/alluxio/master/file/meta/Inode.java]: nit: `oldState` for consistency with `newState`

File: core/server/master/src/main/java/alluxio/master/file/meta/InodeDirectoryIdGenerator.java [hunk 18]
```diff
@@ -41,7 +42,20 @@ public InodeDirectoryIdGenerator(ContainerIdGenerable containerIdGenerator) {
     mContainerIdGenerator = Preconditions.checkNotNull(containerIdGenerator);
   }
 
+  /**
+   * @return the next directory id
+   */
   synchronized long getNewDirectoryId() {
+    return getNewDirectoryId(null);
```
Comment (jsimsa) [inline core/server/master/src/main/java/alluxio/master/file/meta/InodeDirectoryIdGenerator.java]: nit: if you pass `NoopJournalContext.INSTANCE` than the callee can assume the journalContext is never null

File: core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java [hunk 19]
```diff
@@ -558,97 +558,138 @@ public CreatePathResult createPath(LockedInodePath inodePath, CreatePathOptions<
...
-        toPersistDirectories.add(lastInode);
-      } else if (!lastInode.isDirectory() || !(options instanceof CreateDirectoryOptions
-          && ((CreateDirectoryOptions) options).isAllowExists())) {
-        String errorMessage = ExceptionMessage.FILE_ALREADY_EXISTS.getMessage(path);
-        LOG.error(errorMessage);
-        throw new FileAlreadyExistsException(errorMessage);
+    Inode<?> lastInode = null;
+    while (lastInode == null) {
+      // Try to lock the last inode with the lock mode of the path.
+      if (extensibleInodePath.getLockMode() == LockMode.READ) {
```
Comment (jsimsa) [inline core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java]: what do you think about using a switch statement here to make it explicit that we get write-lock for both `LockMode.WRITE` and `LockMode.WRITE_PARENT`?

File: core/server/master/src/main/java/alluxio/master/file/meta/InodeUtils.java [hunk 20]
```diff
@@ -0,0 +1,89 @@
...
+import alluxio.underfs.UnderFileSystem;
+import alluxio.underfs.options.MkdirsOptions;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+
+/**
+ * Represents the tree of Inode's.
```
Comment (jsimsa) [inline core/server/master/src/main/java/alluxio/master/file/meta/InodeUtils.java]: fix javadoc

File: core/server/master/src/main/java/alluxio/master/file/meta/InodeUtils.java [hunk 21]
```diff
@@ -0,0 +1,89 @@
...
+   *
+   * @param dir the {@link InodeDirectory} to persist
+   * @param inodeTree the {@link InodeTree}
+   * @param mountTable the {@link MountTable}
+   * @param journalContext the journal context
+   * @throws IOException if the file fails to persist
+   */
+  public static void syncPersistDirectory(InodeDirectory dir, InodeTree inodeTree,
+      MountTable mountTable, JournalContext journalContext) throws IOException {
+    RetryPolicy retry = new ExponentialBackoffRetry(2, 1000, 200);
```
Comment (jsimsa) [inline core/server/master/src/main/java/alluxio/master/file/meta/InodeUtils.java]: 200 retries is pretty generous

Comment (jsimsa) [inline core/server/master/src/main/java/alluxio/master/file/meta/InodeUtils.java]: do we really want to wait for ~200 seconds before timing out?

File: core/server/master/src/main/java/alluxio/master/file/meta/InodeUtils.java [hunk 22]
```diff
@@ -0,0 +1,89 @@
...
+          dir.setPersistenceState(PersistenceState.PERSISTED);
+
+          // Append the persist entry to the journal.
+          File.PersistDirectoryEntry persistDirectory =
+              File.PersistDirectoryEntry.newBuilder().setId(dir.getId()).build();
+          journalContext.append(
+              Journal.JournalEntry.newBuilder().setPersistDirectory(persistDirectory).build());
+          success = true;
+        } catch (Exception e) {
+          // Ignore
```
Comment (jsimsa) [inline core/server/master/src/main/java/alluxio/master/file/meta/InodeUtils.java]:
```
LOG.warn("Failed to persist directory {} : {}", ufsUri, e.getMessage());
LOG.debug("Exception: ", e);
```

File: core/server/common/src/main/java/alluxio/master/AbstractMaster.java (hunk 7 above)
Comment (gpang) [inline core/server/common/src/main/java/alluxio/master/AbstractMaster.java]: Done. For some reason intellij did not handle this well.

File: core/server/common/src/main/java/alluxio/master/journal/JournalContext.java (hunk 9 above)
Comment (gpang) [inline core/server/common/src/main/java/alluxio/master/journal/JournalContext.java]: Done. For some reason intellij did not handle this well.

File: core/server/common/src/main/java/alluxio/master/journal/JournalContext.java (hunk 8 above)
Comment (gpang) [inline core/server/common/src/main/java/alluxio/master/journal/JournalContext.java]: Done. For some reason intellij did not handle this well.

File: core/server/common/src/main/java/alluxio/master/journal/NoopJournalContext.java (hunk 10 above)
Comment (gpang) [inline core/server/common/src/main/java/alluxio/master/journal/NoopJournalContext.java]: Done. For some reason intellij did not handle this well.

File: core/server/master/src/main/java/alluxio/master/block/BlockMaster.java (hunk 11 above)
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/block/BlockMaster.java]: I moved JournalContext out as an interface, instead of a private inner class.

File: core/server/master/src/main/java/alluxio/master/file/meta/Inode.java (hunk 17 above)
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/file/meta/Inode.java]: Done

File: core/server/master/src/main/java/alluxio/master/file/meta/InodeDirectoryIdGenerator.java (hunk 18 above)
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/file/meta/InodeDirectoryIdGenerator.java]: Good catch! done.

File: core/server/master/src/main/java/alluxio/master/file/meta/InodeUtils.java (hunk 20 above)
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/file/meta/InodeUtils.java]: Done

File: core/server/master/src/main/java/alluxio/master/file/meta/InodeUtils.java (hunk 22 above)
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/file/meta/InodeUtils.java]: Done

File: core/server/master/src/main/java/alluxio/master/file/meta/InodeUtils.java (hunk 21 above)
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/file/meta/InodeUtils.java]: I wanted to wait long enough for slow UFSes. I reduced it to 50 retries, which would be between 40-50 seconds.

File: core/server/master/src/main/java/alluxio/master/file/FileSystemMaster.java (hunk 3 above)
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/file/FileSystemMaster.java]: That is a good question. I do think there is a difference between (B) and (C), because all the `entry` methods will call (C), and not (B). I think (B) can be inlined with (A). That type of refactor is independent and can be done at a later time.

Comment (jsimsa) [inline core/server/master/src/main/java/alluxio/master/file/FileSystemMaster.java]: I do not think it is a completely independent refactor before this PR breaks the invariant that `Journal` methods journal, while `Internal` methods do not (reducing the readability of the code, which is the reason I propose to get rid of the `Journal` methods) ... having said that I agree that it makes more sense to do it in a separate PR

**Comment by jsimsa** [comment]
please resolve merge conflicts with the `master` branch

**Comment by gpang** [comment]
Thanks @jsimsa for the review!

File: core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java [hunk 23]
```diff
@@ -558,97 +558,134 @@ public CreatePathResult createPath(LockedInodePath inodePath, CreatePathOptions<
...
+          lastInode = currentInodeDirectory.getChildReadLock(name, lockList);
+          break;
+        case WRITE_PARENT:
+        case WRITE:
+          lastInode = currentInodeDirectory.getChildWriteLock(name, lockList);
+          break;
+        default:
+          // This should not be reachable.
+          LOG.warn("Unexpected lock mode encountered: {}",
+              extensibleInodePath.getLockMode().toString());
```
Comment (calvinjia) [inline core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java]: `toString` should not be required here

File: core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java [hunk 24]
```diff
@@ -558,97 +558,134 @@ public CreatePathResult createPath(LockedInodePath inodePath, CreatePathOptions<
...
-        lastInode = InodeFile.create(mContainerIdGenerator.getNewContainerId(),
-            currentInodeDirectory.getId(), name, System.currentTimeMillis(), fileOptions);
-        // Lock the created inode before subsequent operations, and add it to the lock group.
-        lockList.lockWriteAndCheckNameAndParent(lastInode, currentInodeDirectory, name);
-        if (currentInodeDirectory.isPinned()) {
-          // Update set of pinned file ids.
-          mPinnedInodeFileIds.add(lastInode.getId());
+      } else {
+        // create the new inode, with a write lock
+        if (options instanceof CreateDirectoryOptions) {
```
Comment (calvinjia) [inline core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java]: Not related to this PR, but is the code overlap large enough to justify switching on creating a file or directory based on the options type instead of having separate methods?

File: core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java [hunk 25]
```diff
@@ -558,97 +558,134 @@ public CreatePathResult createPath(LockedInodePath inodePath, CreatePathOptions<
...
-      mInodes.add(lastInode);
-      currentInodeDirectory.addChild(lastInode);
-      currentInodeDirectory.setLastModificationTimeMs(options.getOperationTimeMs());
-    }
+        if (!currentInodeDirectory.addChild(lastInode)) {
+          // Could not add the child inode to the parent. Continue and try again.
+          lockList.unlockLast();
+          lastInode = null;
+          continue;
+        }
```
Comment (calvinjia) [inline core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java]: Is there a general guideline to follow when aborting and trying again in terms of what state needs to be reset?

File: core/server/master/src/main/java/alluxio/master/file/meta/InodeUtils.java [hunk 26]
```diff
@@ -0,0 +1,102 @@
...
+   * Synchronously persists an {@link InodeDirectory} to the UFS. If concurrent calls are made, only
+   * one thread will persist to UFS, and the others will wait until it is persisted.
+   *
+   * @param dir the {@link InodeDirectory} to persist
+   * @param inodeTree the {@link InodeTree}
+   * @param mountTable the {@link MountTable}
+   * @param journalContext the journal context
+   * @throws IOException if the file fails to persist
+   */
+  public static void syncPersistDirectory(InodeDirectory dir, InodeTree inodeTree,
```
Comment (calvinjia) [inline core/server/master/src/main/java/alluxio/master/file/meta/InodeUtils.java]: This does not feel like a utility method. Is it possible to eventually have this belong to the file system master?

File: core/server/master/src/main/java/alluxio/master/file/meta/InodeUtils.java [hunk 27]
```diff
@@ -0,0 +1,102 @@
...
+        try {
+          AlluxioURI uri = inodeTree.getPath(dir);
+          MountTable.Resolution resolution = mountTable.resolve(uri);
+          String ufsUri = resolution.getUri().toString();
+          errorDetails.append(" ufsUri: ").append(ufsUri);
+          UnderFileSystem ufs = resolution.getUfs();
+          MkdirsOptions mkdirsOptions =
+              MkdirsOptions.defaults().setCreateParent(false).setOwner(dir.getOwner())
+                  .setGroup(dir.getGroup()).setMode(new Mode(dir.getMode()));
+          if (!ufs.isDirectory(ufsUri)) {
```
Comment (calvinjia) [inline core/server/master/src/main/java/alluxio/master/file/meta/InodeUtils.java]: Does this check guarantee anything?

File: core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java [hunk 28]
```diff
@@ -608,7 +611,7 @@ public StartupConsistencyCheck getStartupConsistencyCheck() {
   @Override
   public long getFileId(AlluxioURI path) throws AccessControlException {
     try (JournalContext journalContext = createJournalContext();
-        LockedInodePath inodePath = mInodeTree.lockInodePath(path, InodeTree.LockMode.WRITE)) {
+         LockedInodePath inodePath = mInodeTree.lockInodePath(path, InodeTree.LockMode.WRITE)) {
```
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java]: I don't know. I think there are also others in this file.

File: core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java [hunk 29]
```diff
@@ -1623,12 +1588,13 @@ private void renameAndJournal(LockedInodePath srcInodePath, LockedInodePath dstI
...
    * @param replayed whether the operation is a result of replaying the journal
    * @param options method options
+   * @param journalContext the journal context
    * @throws FileDoesNotExistException if a non-existent file is encountered
    * @throws InvalidPathException if an invalid path is encountered
    * @throws IOException if an I/O error is encountered
    */
   private void renameInternal(LockedInodePath srcInodePath, LockedInodePath dstInodePath,
-      boolean replayed, RenameOptions options)
+      boolean replayed, RenameOptions options, JournalContext journalContext)
```
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java]: The current ordering is for output parameters to follow input parameters.

File: core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java (hunk 23 above)
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java]: Removed.

File: core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java (hunk 24 above)
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java]: I think this is the only part of this method which needs to know which type of inode to create for the leaf. The rest of this method is code overlap. Are you suggesting to convert this method to `createFilePath` and `createDirPath`?

File: core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java (hunk 25 above)
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java]: I updated the code a little bit.

The guideline is to update the state after a successful add, so cleanup is not necessary.

File: core/server/master/src/main/java/alluxio/master/file/meta/InodeUtils.java (hunk 27 above)
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/file/meta/InodeUtils.java]: This was what the previous "persist" codepath did. If the dir already exists, it will skip the `mkdirs`.

This will change if we update the UFS api.

File: core/server/master/src/main/java/alluxio/master/file/meta/InodeUtils.java (hunk 26 above)
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/file/meta/InodeUtils.java]: What does it mean that this is not a utility method?

Is the suggestion to move this static method to `FileSystemMaster.java`?

File: core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java (hunk 24 above)
Comment (calvinjia) [inline core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java]: Yeah, if the code overlap is fairly isolated, we could have another method which handles the overlap.

File: core/server/master/src/main/java/alluxio/master/file/meta/InodeUtils.java (hunk 26 above)
Comment (calvinjia) [inline core/server/master/src/main/java/alluxio/master/file/meta/InodeUtils.java]: Yeah, it does not seem like an operation any other portion of the code should execute (though currently it is being used in places like BlockMaster)

File: core/server/master/src/main/java/alluxio/master/file/meta/InodeUtils.java (hunk 27 above)
Comment (calvinjia) [inline core/server/master/src/main/java/alluxio/master/file/meta/InodeUtils.java]: Perhaps we should just call `mkdirs` and catch the appropriate exception? There is no guarantee that the directory does not exist after the call anyway, so if we hit an error it should be handled.

**Comment by calvinjia** [comment]
@gpang Thanks! I left a few follow up comments

File: core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java (hunk 24 above)
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java]: I added a TODO to refactor this.

File: core/server/master/src/main/java/alluxio/master/file/meta/InodeUtils.java (hunk 26 above)
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/file/meta/InodeUtils.java]: Sounds good. I moved this functionality into InodeTree.

File: core/server/master/src/main/java/alluxio/master/file/meta/InodeUtils.java (hunk 27 above)
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/file/meta/InodeUtils.java]: I updated this to simply throw the exceptions encountered.

**Comment by gpang** [comment]
@calvinjia Thanks for the review. I updated the PR.
